File: java/03-advanced/05-data-structures/StackQueueDemo.java
import java.util.*;

/**
 * LESSON: Stack, Queue, Deque, PriorityQueue
 * Fundamental data structures for organizing data.
 */
public class StackQueueDemo {
    public static void main(String[] args) {
        // ===== STACK (LIFO - Last In, First Out) =====
        System.out.println("--- Stack (LIFO) ---");
        Stack<String> stack = new Stack<>();
        stack.push("A");
        stack.push("B");
        stack.push("C");
        System.out.println("Stack: " + stack);
        System.out.println("Peek (top): " + stack.peek());     // C (doesn't remove)
        System.out.println("Pop: " + stack.pop());              // C (removes)
        System.out.println("Pop: " + stack.pop());              // B
        System.out.println("After pops: " + stack);             // [A]
        System.out.println("Empty? " + stack.isEmpty());

        // ===== QUEUE (FIFO - First In, First Out) =====
        System.out.println("\n--- Queue (FIFO) ---");
        Queue<String> queue = new LinkedList<>();
        queue.offer("First");   // add to end
        queue.offer("Second");
        queue.offer("Third");
        System.out.println("Queue: " + queue);
        System.out.println("Peek (front): " + queue.peek());   // First
        System.out.println("Poll: " + queue.poll());            // First (removes from front)
        System.out.println("Poll: " + queue.poll());            // Second
        System.out.println("After polls: " + queue);            // [Third]

        // ===== DEQUE (Double-Ended Queue) =====
        System.out.println("\n--- Deque (Double-Ended) ---");
        Deque<String> deque = new ArrayDeque<>();
        deque.addFirst("B");
        deque.addFirst("A");   // add to front
        deque.addLast("C");    // add to back
        deque.addLast("D");
        System.out.println("Deque: " + deque);          // [A, B, C, D]
        System.out.println("First: " + deque.peekFirst());
        System.out.println("Last: " + deque.peekLast());
        deque.removeFirst();
        deque.removeLast();
        System.out.println("After removing ends: " + deque); // [B, C]

        // Using Deque as Stack (preferred over Stack class)
        System.out.println("\n--- Deque as Stack ---");
        Deque<Integer> dequeStack = new ArrayDeque<>();
        dequeStack.push(10);
        dequeStack.push(20);
        dequeStack.push(30);
        System.out.println("Push 10, 20, 30: " + dequeStack);
        System.out.println("Pop: " + dequeStack.pop()); // 30

        // ===== PRIORITY QUEUE (elements ordered by priority) =====
        System.out.println("\n--- PriorityQueue (Min-Heap) ---");
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        minHeap.offer(30);
        minHeap.offer(10);
        minHeap.offer(20);
        minHeap.offer(5);
        System.out.println("PQ: " + minHeap); // internal order may vary
        System.out.print("Poll order (min first): ");
        while (!minHeap.isEmpty()) {
            System.out.print(minHeap.poll() + " "); // 5 10 20 30
        }
        System.out.println();

        // Max-Heap (reverse order)
        System.out.println("\n--- PriorityQueue (Max-Heap) ---");
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
        maxHeap.offer(30);
        maxHeap.offer(10);
        maxHeap.offer(20);
        maxHeap.offer(5);
        System.out.print("Poll order (max first): ");
        while (!maxHeap.isEmpty()) {
            System.out.print(maxHeap.poll() + " "); // 30 20 10 5
        }
        System.out.println();
    }
}

File: cpp-practice/03-advanced/04-data-structures/stack_queue_demo.cpp
/**
 * LESSON: Stack, Queue, Deque, Priority Queue
 */
#include <iostream>
#include <stack>
#include <queue>
#include <deque>
#include <vector>
#include <functional> // std::greater

int main() {
    // ===== STACK (LIFO) =====
    std::cout << "--- Stack (LIFO) ---\n";
    std::stack<std::string> stack;
    stack.push("A");
    stack.push("B");
    stack.push("C");
    std::cout << "Top: " << stack.top() << "\n"; // C
    std::cout << "Size: " << stack.size() << "\n";

    std::cout << "Pop order: ";
    while (!stack.empty()) {
        std::cout << stack.top() << " "; // C B A
        stack.pop();
    }
    std::cout << "\n";

    // ===== QUEUE (FIFO) =====
    std::cout << "\n--- Queue (FIFO) ---\n";
    std::queue<std::string> queue;
    queue.push("First");
    queue.push("Second");
    queue.push("Third");
    std::cout << "Front: " << queue.front() << "\n"; // First
    std::cout << "Back: " << queue.back() << "\n";   // Third

    std::cout << "Pop order: ";
    while (!queue.empty()) {
        std::cout << queue.front() << " "; // First Second Third
        queue.pop();
    }
    std::cout << "\n";

    // ===== DEQUE (Double-Ended Queue) =====
    std::cout << "\n--- Deque ---\n";
    std::deque<int> dq;
    dq.push_back(2);
    dq.push_back(3);
    dq.push_front(1);
    dq.push_back(4);
    dq.push_front(0);

    std::cout << "Deque: ";
    for (int n : dq) std::cout << n << " "; // 0 1 2 3 4
    std::cout << "\n";

    dq.pop_front();
    dq.pop_back();
    std::cout << "After pops: ";
    for (int n : dq) std::cout << n << " "; // 1 2 3
    std::cout << "\n";

    // ===== PRIORITY QUEUE (Max-Heap by default) =====
    std::cout << "\n--- Priority Queue (Max-Heap) ---\n";
    std::priority_queue<int> maxHeap;
    maxHeap.push(30);
    maxHeap.push(10);
    maxHeap.push(20);
    maxHeap.push(5);

    std::cout << "Pop order (max first): ";
    while (!maxHeap.empty()) {
        std::cout << maxHeap.top() << " "; // 30 20 10 5
        maxHeap.pop();
    }
    std::cout << "\n";

    // Min-Heap
    std::cout << "\n--- Priority Queue (Min-Heap) ---\n";
    std::priority_queue<int, std::vector<int>, std::greater<int>> minHeap;
    minHeap.push(30);
    minHeap.push(10);
    minHeap.push(20);
    minHeap.push(5);

    std::cout << "Pop order (min first): ";
    while (!minHeap.empty()) {
        std::cout << minHeap.top() << " "; // 5 10 20 30
        minHeap.pop();
    }
    std::cout << "\n";

    return 0;
}
